File: 2017/25/IsaacG.py
#!/bin/python

import re
from lib import aoc
InputType = tuple[str, int, dict[str, list[tuple[bool, int, str]]]]


def solver(puzzle_input: InputType, part_one: bool) -> int:
    cursor = 0
    state, steps, rules = puzzle_input
    tape = set()
    for _ in range(steps):
        write, move, state = rules[state][cursor in tape]
        if write:
            tape.add(cursor)
        else:
            tape.discard(cursor)
        cursor += move
    return len(tape)

def input_parser(puzzle_input: str) -> InputType:
    """Parse the input data."""
    preamble, *rules = puzzle_input.split("\n\n")
    patt = r"Begin in state (.)\.\nPerform a diagnostic checksum after (\d+) steps."
    m = re.match(patt, preamble)
    assert m
    initial_state = m.group(1)
    steps = int(m.group(2))

    patt = re.compile(r"""In state (.):
If the current value is 0:
- Write the value ([01])\.
- Move one slot to the (left|right)\.
- Continue with state (.)\.
If the current value is 1:
- Write the value ([01])\.
- Move one slot to the (left|right)\.
- Continue with state (.)\.""")

    res = {}
    for rule in rules:
        m = patt.match(rule)
        assert m
        start, w0, m0, n0, w1, m1, n1 = m.groups()
        res[start] = [
                (w0 == "1", 1 if m0 == "right" else -1, n0),
                (w1 == "1", 1 if m1 == "right" else -1, n1),
        ]
    return initial_state, steps, res

File: 2017/25/IsaacG.go
package y2017

import (
	"regexp"
	"strings"
	"isaacgood.com/aoc/helpers"
)

var preambleRe = regexp.MustCompile(`Begin in state (.)\.\nPerform a diagnostic checksum after ([0-9]+) steps.`)
var ruleRe = regexp.MustCompile(`In state (.):
  If the current value is 0:
    - Write the value ([01])\.
    - Move one slot to the (left|right)\.
    - Continue with state (.)\.
  If the current value is 1:
    - Write the value ([01])\.
    - Move one slot to the (left|right)\.
    - Continue with state (.)\.`)

type turingRule struct {
	write bool
	move int
	nextState string
}

func newTuringRule(parts []string) *turingRule {
	return &turingRule{
		write: parts[0] == "1",
		move: map[string]int{"left": -1, "right": 1}[parts[1]],
		nextState: parts[2],
	}
}

// Day25 solves 2017/25.
type Day25 struct {
	rules map[string][2]*turingRule
	initialState string
	steps int
}

// New25 returns a new solver for 2017/25.
func New25() *Day25 {
	return &Day25{}
}

// SetInput handles input for this solver.
func (p *Day25) SetInput(data string) {
	chunks := strings.Split(data, "\n\n")
	matches := preambleRe.FindStringSubmatch(chunks[0])
	p.initialState = matches[1]
	p.steps = helpers.Atoi(matches[2])
	p.rules = make(map[string][2]*turingRule)
	for _, chunk := range chunks[1:] {
		matches = ruleRe.FindStringSubmatch(chunk)
		p.rules[matches[1]] = [2]*turingRule{
			newTuringRule(matches[2:5]),
			newTuringRule(matches[5:8]),
		}
	}
}

// Solve returns the solution for one part.
func (p *Day25) Solve(part int) string {
	state := p.initialState
	cursor := 0
	tape := make(map[int]interface{})
	for range(p.steps) {
		var idx int
		if _, ok := tape[cursor]; ok {
			idx = 1
		} else {
			idx = 0
		}
		rule := p.rules[state][idx]
		if rule.write {
			tape[cursor] = struct{}{}
		} else {
			delete(tape, cursor)
		}
		cursor += rule.move
		state = rule.nextState
	}
        return helpers.Itoa(len(tape))
}
